### PR TITLE
Mining on user devices (#11)

### LINKED ISSUE #11: Mining on user devices
**Description:**
The BisonCoin app will need to support mining on user  devices, which will mean the front-end will execute the mining code on compute the proofOfWork for a transaction  that is needing to be mined

**Acceptance Criteria**
- [x] Users can mine, the mining work is done on user devices
- [x] Users get rewarded when they participate in mining
- [x] Implement logic to turn mining on and off based on the component toggle, including connecting and disconnecting the client from the socket
- [x] Implement proofOfWork algorithm
- [x] Implement logic to receive messages from the socket and trigger mining, cease mining, and reward logic for all the paths
- [x] Ensure mining-service coordinates the mining session as intended

### PR DESCRIPTION
**Pull request by ZackHolmberg**
closes #11

### CODE REVIEW

File: src/backend/services/mining-api/src/routes.py [hunk 1]
```diff
@@ -41,72 +43,94 @@ def send_to_connected_clients(transaction):
...
 def handle(message):
     socketio.emit("response", message)
 
+
 @socketio.on('proof')
 def handle_proofs(message):
-    global ongoing_proof, ongoing_transaction, blockchain_url, blockchain_wallet_url, transactions
-    
+    global ongoing_proof, ongoing_transaction, blockchain_url, transactions
+    # print("ZACK OUTPUT - GETTING HERE 0", file=sys.stderr)
```
Comment (ZackHolmberg) [inline src/backend/services/mining-api/src/routes.py]: Leaving these print statements here for now until the multiple findProof trigger bug is resolved

File: src/frontend/src/App.vue [hunk 2]
```diff
@@ -1,13 +1,102 @@
...
+  }
+
+  proofOfWork(transaction: Transaction): any  {
+      let nonce = -1;
+      let hash = "";
+      while (!this.validHash(hash) && this.findProof) {
+        nonce += 1;
+        hash = this.computeHash(nonce, transaction);
+      }
+      console.log("Returning: ",{proof: hash, nonce: nonce})
```
Comment (ZackHolmberg) [inline src/frontend/src/App.vue]: Leaving these print statements here for now until the multiple findProof trigger bug is resolved

File: src/frontend/src/pages/NewTransactionPage.vue [hunk 3]
```diff
@@ -49,7 +49,11 @@ import Button from "../components/Button.vue";
 })
 export default class NewTransactionPage extends Vue {
   newTransaction() {
-    const values = { amount: 10, contact: "email" };
+    const values = {
+      amount: 1,
+      contact:
+        "-----BEGIN PUBLIC KEY-----\r\nMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCnSXKzVWvTV/wSiG9Rimapz0eA\r\nza9atWCXnU/hTXvOdHpxCpejzOUsB5hPbFC7EdUj4ffHnj/IPYIOgMNzX+i8SwHa\r\nz+gl0K7n8JUzufRnom4/WMHMpkvqUYnVT9/TQLOMyBd/uEEh7is7ioX8EOFMevDu\r\n98aZ4ouHDO9YTNlH4QIDAQAB\r\n-----END PUBLIC KEY-----\r\n",
```
Comment (ZackHolmberg) [inline src/frontend/src/pages/NewTransactionPage.vue]: TODO: Change this from a hardcoded value to accepting the input passed in the input field

File: src/frontend/src/App.vue [hunk 4]
```diff
@@ -1,13 +1,102 @@
...
+          }
+
+          console.log("Sending: ",toSend)
+          socket.emit("proof",toSend)
+          this.$store.commit('MUTATION_SET_FIND_PROOF', false);
+
+        }
+      });
+
+      socket.on("reward", () => {
```
Comment (ZackHolmberg) [inline src/frontend/src/App.vue:82]: TODO:

1. Move displaying toast into an action
2. Show toasts for each one of these paths: findProof, reward and stopProof

File: src/frontend/src/store/index.ts [hunk 5]
```diff
@@ -367,26 +397,40 @@ export default new Vuex.Store({
...
-          "public_key": getters.walletId,
+          "public_key": walletId,
           "password": password,
         })
         .then(
           () => {
-            commit("MUTATATION_SET_LOADING", false);
+            commit("MUTATION_SET_LOADING", false);
+
+            // TODO: Uncomment when we read in and set privateKey on login
```
Comment (ZackHolmberg) [inline src/frontend/src/store/index.ts:372]: TODO: Create an issue to tackle reading and writing privateKeys from local text file. Thought it was gonna be obvious but it isn't!

File: src/backend/services/users-api/src/app/routes.py [hunk 6]
```diff
@@ -22,11 +20,27 @@ def login():
...
+        user = mongo.db.users.find_one({"umnetID": umnetID})
+        correct_credentials = user["password"] == password
+    except Exception as e:
+        raise UserNotFoundException()
 
-    return jsonify(success=(user["password"] == password))
+    if not correct_credentials:
+        return jsonify(error="Incorrect password for corresponding UMnetId"), HttpCode.BAD_REQUEST.value
+        # raise IncorrectCredentialsException() >>> I want to do something like this but
+        # not sure the best way to do it! Suggestions please!
```
Comment (abhisachdev17) [inline src/backend/services/users-api/src/app/routes.py]: That might be the best way to do it, so it gets caught in the error handlers

File: src/frontend/src/App.vue [hunk 7]
```diff
@@ -1,13 +1,102 @@
...
 </template>
 
 <script lang="ts">
 import { Vue } from "vue-property-decorator";
+import { io } from "socket.io-client";
+import { sha256 } from "js-sha256";
+import type { Transaction } from "./types";
 
-export default class App extends Vue {}
+const socket = io("http://localhost");
```
Comment (madisonfines) [inline src/frontend/src/App.vue]: You can remove some of these extra lines and maybe write a comment above socket to have a defining area for variables outside of export default!

File: src/frontend/src/App.vue [hunk 8]
```diff
@@ -1,13 +1,102 @@
...
+  get findProof(){
+    return this.$store.getters.findProof
+  }
+
+  validHash (hash: string): boolean  {
+      return hash.startsWith("0000", 0);
+  }
+
+  computeHash(nonce: number, transaction: Transaction): string {
+
```
Comment (madisonfines) [inline src/frontend/src/App.vue]: Remove space

File: src/frontend/src/App.vue [hunk 9]
```diff
@@ -1,13 +1,102 @@
...
+      const toHash = (nonce.toString()+transaction.amount.toString()+transaction.id+transaction.signature).replace(/(\r\n|\n|\r)/gm, "");
```
Comment (madisonfines) [inline src/frontend/src/App.vue]: Could you maybe indent this or add space between pluses so it is easier to read please!

File: src/frontend/src/components/Mining.vue [hunk 10]
```diff
@@ -36,7 +36,7 @@ export default class Mining extends Vue {
   }
 
   onChange() {
-      this.$store.commit('MUTATATION_SET_MINING', this.$data.miningValue);
+      this.$store.commit('MUTATION_SET_MINING', this.$data.miningValue);
```
Comment (madisonfines) [inline src/frontend/src/components/Mining.vue]: Still can't believe I did this

**Review by madisonfines — CHANGES_REQUESTED** [review]
Eventually we will have to write some e2e tests, it just might be a lot of tests so I'm not sure if it would be a better idea to create a separate PR for that

File: src/frontend/src/components/Mining.vue (hunk 10 above)
Comment (ZackHolmberg) [reply src/frontend/src/components/Mining.vue]: gif

**Review by madisonfines — APPROVED** [review]

**Review by abhisachdev17 — APPROVED** [review]
damn, that's some nice code.

File: src/frontend/src/pages/CreateAccountPage.vue [hunk 11]
```diff
@@ -39,6 +39,8 @@
       id="create-account-button"
       dest=""
       label="Create Account"
+      size="small"
+      type="default"
```
Comment (abhisachdev17) [inline src/frontend/src/pages/CreateAccountPage.vue:45]: can you add the password input type? should be a quick fix

Comment (ZackHolmberg) [reply src/frontend/src/pages/CreateAccountPage.vue:45]: Forgot about that! Fixed

**Pull request merged**